### PR TITLE
Update to latest Kerberos library

### PR DESCRIPTION
**Pull request by tyrannosaurus-becks**
The version of the `gokrb5` library in this repo is from July 2018, and it's not easy to construct a valid login. There has also been a major refactor of `gokrb5` since then.

The current version of the `gokrb5` documents how to create a valiid SPNEGO login, which we can use for building a CLI handler. Even if we need edits to the `gokrb5` library, we can submit them to master and then pull them in when they're merged. This makes much more sense than PR'ing to a much older version of their library.

This PR updates our code to _a fork of_ the latest version of the `gokrb5` library. This fork is necessary for successful logins until https://github.com/jcmturner/gokrb5/pull/331 is merged.

Please note that our build runs an end-to-end integration test which is unchanged, and it verifies that an identical login can be performed both before this change (on master), and after it (with the updated code), showing that the plugin's external API remains unchanged.

### CODE REVIEW

File: path_config.go [hunk 1]
```diff
@@ -53,7 +51,7 @@ func (b *backend) pathConfigRead(ctx context.Context, req *logical.Request, data
 	} else {
 		return &logical.Response{
 			Data: map[string]interface{}{
-				"keytab":          config.Keytab,
+				// keytab is intentionally not returned here because it's sensitive
```
Comment (kalafut) [inline path_config.go]: Consider also marking in Sensitive in the FieldSchema.

File: path_config.go [hunk 2]
```diff
@@ -72,11 +70,7 @@ func (b *backend) pathConfigWrite(ctx context.Context, req *logical.Request, dat
 	}
 
 	// Check that the keytab is valid by parsing with krb5go
-	binary, err := base64.StdEncoding.DecodeString(kt)
-	if err != nil {
-		return logical.ErrorResponse(fmt.Sprintf("could not base64 decode keytab: %v", err)), logical.ErrInvalidRequest
-	}
-	if _, err = keytab.Parse(binary); err != nil {
+	if _, err := parseKeytab(kt); err != nil {
 		return logical.ErrorResponse(fmt.Sprintf("invalid keytab: %v", err)), logical.ErrInvalidRequest
```
Comment (kalafut) [inline path_config.go]:
```suggestion
		return logical.ErrorResponse("invalid keytab: %v", err), logical.ErrInvalidRequest
```

File: path_login.go [hunk 3]
```diff
@@ -91,20 +65,15 @@ func (b *backend) pathLoginGet(ctx context.Context, req *logical.Request, d *fra
 func (b *backend) pathLoginUpdate(ctx context.Context, req *logical.Request, d *framework.FieldData) (*logical.Response, error) {
 	kerbCfg, err := b.config(ctx, req.Storage)
 	if err != nil {
-		return nil, err
+		return nil, errwrap.Wrap(errors.New("unable to get kerberos config"), err)
```
Comment (kalafut) [inline path_login.go]:
```suggestion
		return nil, errwrap.Wrap("unable to get kerberos config: {{err}}", err)
```

Comment (kalafut) [reply path_login.go]: Same question for some other instances in this file.

Comment (tyrannosaurus-becks) [reply path_login.go]: Will update! It's been a while since I used that library.

File: path_config.go (hunk 1 above)
Comment (tyrannosaurus-becks) [reply path_config.go]: @kalafut actually, it looks like it's not possible unless I bump the version of the Vault library and update all those vendored dependencies. It's currently using v0.1.13 of the Vault SDK, when "sensitive" didn't exist yet. Would you mind if I instead put it into the backlog TODO?

File: path_login.go (hunk 3 above)
Comment (tyrannosaurus-becks) [reply path_login.go]: Addressed in https://github.com/hashicorp/vault-plugin-auth-kerberos/pull/19/commits/b93838f9697086f98194c5b0d210462d72fe17c9.

File: path_config.go (hunk 1 above)
Comment (kalafut) [reply path_config.go]: I don't mind.

Comment (tyrannosaurus-becks) [reply path_config.go]: Done!

File: path_login.go [hunk 4]
```diff
@@ -234,3 +250,25 @@ func (b *backend) pathLoginUpdate(ctx context.Context, req *logical.Request, d *
...
+func (w *simpleResponseWriter) Write(b []byte) (int, error) {
+	w.body = b
+	return 0, nil
+}
+
+func (w *simpleResponseWriter) WriteHeader(statusCode int) {
+	w.statusCode = statusCode
+}
+
+type simpleServiceSettings struct {
```
Comment (kalafut) [inline path_login.go]: Dead code, I think.

File: path_login.go [hunk 5]
```diff
@@ -234,3 +250,25 @@ func (b *backend) pathLoginUpdate(ctx context.Context, req *logical.Request, d *
...
 	}, nil
 }
+
+type simpleResponseWriter struct {
+	body       []byte
+	statusCode int
+}
+
+func (w *simpleResponseWriter) Header() http.Header {
+	return make(map[string][]string)
```
Comment (kalafut) [inline path_login.go]:
```suggestion
	return make(http.Header)
```

**Pull request merged**